tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/Inliner.scala

```diff
@@ -0,0 +1,238 @@
+    // We fix that below with type unification.
+    val freshBody = deepCopy(decl.body)
+
+    // Each formal parameter gets instantiated independently.
+    val replacedBody = decl.formalParams.zip(args).foldLeft(freshBody) { case (partialBody, (fParam, arg)) =>
+      ReplaceFixed(tracker)(NameEx(fParam.name)(arg.typeTag), arg)(partialBody)
+    }
+
+    // There are two cases where the above instantiation might be incomplete:
+    // a) In the case of an application of the form A(B()), `arg` will have the value `B()`, which is _not_
+    //    a fully inlined expression
+    // b) In the case of an application of the form A(B), where A is a HO operator and `B` is an operator name
+    //    Then, if A(F(_)) == e, e[B/F] would contain applications of the form B(...), which are _not_ fully inlined
+    // To cover both cases at once, we run an additional transform on the replaced body
+    val newBody = transform(scope)(replacedBody)
+
+    // If the operator has a parametric signature, we have to substitute type parameters with concrete parameters
+    // 1. Unify the operator type with the arguments.
+    // 2. Apply the resulting substitution to the types in all subexpressions.
+    val actualType = OperT1(args.map(_.typeTag.asTlaType1()), freshBody.typeTag.asTlaType1())
+
+    val (substitution, _) = getSubstitution(actualType, decl)
+
+    if (substitution.isEmpty) newBody
+    else new TypeSubstitutor(tracker, substitution)(newBody)
+  }
+
+  // Assume name is in scope. Creates a local LET-IN for call-by-name operators.
+  private def embedCallByName(scope: Scope)(nameEx: NameEx): TlaEx = {
+    val decl = scope(nameEx.name)
+
+    // like in instantiateWithArgs, we compare the declaration type to the expected monotype
+    val freshBody = deepCopy(decl.body)
+    val monoOperType = nameEx.typeTag.asTlaType1()
+
+    val (substitution, tp) = getSubstitution(monoOperType, decl)
+
+    val tpTag = Typed(tp)
+
+    val newBody =
+      if (substitution.isEmpty) freshBody
+      else new TypeSubstitutor(tracker, substitution)(freshBody)
+
+    // To make a local definition, we use a fresh name
+    val newName = nameGenerator.newName()
+    val newLocalDecl = TlaOperDecl(newName, decl.formalParams, newBody)(tpTag)
+
+    LetInEx(NameEx(newName)(tpTag), newLocalDecl)(tpTag)
+  }
+
+  // main access method, performs the transformations described above
+  def transform(scope: Scope): TlaExTransformation = tracker.trackEx {
+    // Standard application
+    case OperEx(TlaOper.apply, NameEx(name), args @ _*) if scope.contains(name) =>
+      instantiateWithArgs(scope)(name, args)
+
+    // Roundabout application, happens if you pass a lambda to a HO operator
+    case OperEx(TlaOper.apply, LetInEx(NameEx(name), decl @ TlaOperDecl(declName, params, _)), args @ _*)
+        if name == declName && params.size == args.size =>
+      // not 100% optimal, but much easier to read/maintain: just push the lambda into scope and apply the std case
```

Comment:
Why not optimal, what could be improved?

Reply:
Technically you could just do the part that's under
```scala
case opDecl: TlaOperDecl => ...
```
and ignore the `List[TlaDecl]` accumulation, but that's a reeeeeeally minor performance benefit.

Reply:
I'd remove the `not 100% optimal...` then (or explain it).